students/Pirouz_N/lesson04/mailroom.py

```diff
@@ -0,0 +1,185 @@
+    # Printing table
+    print('{:<30.30s}\t|{:^16.16s}\t|{:^12.12s}\t|{:^16.16s}'
+          .format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift'))
+    print('------------------------------------------------------------------------------------------------')
+
+    # printing donor table
+    for donor in donor_db:
+        # printing donor table
+        print(
+            '{:<30.30s}\t ${:>15.15s}\t{:>12.12s}\t ${:>15.15s}'
+            .format(donor, format(float(sum(donor_db[donor])), '.2f'), str(len(donor_db[donor])),
+                    format(float(sum(donor_db[donor]) / len(donor_db[donor])), '.2f')))
+    print('\n')
+
+
+def print_list_of_donors():
+    """"This function will print list of donors."""
+    print('{:<30.30s}\t'.format('Donor Name'))
+    print('---------------------------------')
+
+    # printing donor names
+    for donor in donor_db:
+        # printing donor name
+        print('{:<30.30s}\t'.format(donor))
+    print('\n')
+
+
+def write_thank_you(donername):
+    """"This function will write a thank to a donor.
+
+    Args:
+        donername: Is the name of the donor thank you card will be written to.
+
+    """
+    letter = 'Dear {},\n\n'.format(donername)
+    letter += 'Thank you for your latest donation of ${:0.2f}. With this donation your overall donation has reached'\
+        .format(donor_db[donername][len(donor_db[donername]) - 1])
+    letter += ' ${:0.2f}. We are very grateful for all your {} donations, and we appreciate all your support. '\
+        .format(float(sum(donor_db[donername])), len(donor_db[donername]))
+    letter += 'Please do not forget us in future we need because there is still more work that needs to be done and'
+    letter += ' we need your help to accomplish them.\n\n'
+    letter += 'Best Regards\n'
+    letter += 'Pirouz Naghavi'
+    return letter
+
+
+def send_thank_you_after_donation():
+    """"This function will type a thank card to the selected donor after a donation."""
+    while True:
+        option = input('Please type in the name of the donor you wish to thank.\n'
+                       'If you wish to see the list of all the donors, please type list.\n'
+                       'If you wish to add a new donor to the list, please type in their name.\n'
+                       'If you wish to return to the main menu please type exit and hit enter.\n'
+                       'Please type in your desired option and hit enter:\n')
+
+        if option == 'list':
+            print_list_of_donors()
+
+        elif option == 'exit':
+            break
+
+        else:
+            donor_names_list = [item for item in donor_db]
+            if option in donor_names_list:
+                write_thank_you_with_donation(option)
+
+            else:
+                print('The new donor name you entered will be added to the donor table.')
+                donor_db[option] = []
+                write_thank_you_with_donation(option)
```

Comment:
missing a feature where donation is added for existing donor.

Reply:
I am sorry if write_thank_you_with_donation is not named properly but it does that. It adds new donation to the donation history of the user. I will rename it to ask_donation_write_thank_you. Sorry about the confusion.

Best Regards
Pirouz Naghavi